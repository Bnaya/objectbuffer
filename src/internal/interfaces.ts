import type { Heap } from "../structsGenerator/consts";
import type { TransactionalAllocator } from "./TransactionalAllocator";

export interface MemoryStats {
  available: number;
  used: number;
  total: number;
  top: number;
}

export interface GlobalCarrier {
  allocator: TransactionalAllocator;
  heap: Heap;
}

export interface ExternalArgs {
  /**
   * The load factor of the internal hash maps in use
   * @default 0.75
   */
  readonly hashMapLoadFactor: number;
  /**
   * Minimal initial buckets count of internal hash maps in use
   * @default 8
   */
  readonly hashMapMinInitialCapacity: number;
  /**
   * Allocate additional memory for array pointers,
   * will prevent the reallocation and copy when array is getting bigger
   * @default 0
   */
  readonly arrayAdditionalAllocation: number;
}

export interface ObjectBufferSettings {
  /**
   * The load factor of the internal hash maps in use
   * @default 0.75
   */
  readonly hashMapLoadFactor?: number;
  /**
   * Minimal initial buckets count of internal hash maps in use
   * @default 8
   */
  readonly hashMapMinInitialCapacity?: number;
  /**
   * Allocate additional memory for array pointers,
   * will prevent the reallocation and copy when array is getting bigger
   * @default 0
   */
  readonly arrayAdditionalAllocation?: number;
}

export interface InternalAPI {
  getExternalArgs(): ExternalArgs;
  getCarrier(): Readonly<GlobalCarrier>;
  replaceCarrierContent(carrier: GlobalCarrier): void;
  getEntryPointer(): number;
  destroy(): number;
}
