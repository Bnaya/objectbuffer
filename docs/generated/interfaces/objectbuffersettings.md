[@bnaya/objectbuffer](../README.md) / ObjectBufferSettings

# Interface: ObjectBufferSettings

## Table of contents

### Properties

- [arrayAdditionalAllocation](objectbuffersettings.md#arrayadditionalallocation)
- [hashMapLoadFactor](objectbuffersettings.md#hashmaploadfactor)
- [hashMapMinInitialCapacity](objectbuffersettings.md#hashmapmininitialcapacity)

## Properties

### arrayAdditionalAllocation

• `Optional` `Readonly` **arrayAdditionalAllocation**: `number`

Allocate additional memory for array pointers,
will prevent the reallocation and copy when array is getting bigger

**`default`** 0

#### Defined in

[internal/interfaces.ts:51](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L51)

___

### hashMapLoadFactor

• `Optional` `Readonly` **hashMapLoadFactor**: `number`

The load factor of the internal hash maps in use

**`default`** 0.75

#### Defined in

[internal/interfaces.ts:40](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L40)

___

### hashMapMinInitialCapacity

• `Optional` `Readonly` **hashMapMinInitialCapacity**: `number`

Minimal initial buckets count of internal hash maps in use

**`default`** 8

#### Defined in

[internal/interfaces.ts:45](https://github.com/Bnaya/objectbuffer/blob/266152d/src/internal/interfaces.ts#L45)
