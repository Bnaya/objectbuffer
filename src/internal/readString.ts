import { Heap } from "../structsGenerator/consts";
import { stringDecode } from "./stringDecode";
import {
  string_bytesLength_get,
  string_charsPointer_get,
} from "./generatedStructs";

export function readString(heap: Heap, stringEntryPointer: number) {
  return stringDecode(
    heap.Uint8Array,
    string_charsPointer_get(heap, stringEntryPointer),
    string_bytesLength_get(heap, stringEntryPointer)
  );
}
