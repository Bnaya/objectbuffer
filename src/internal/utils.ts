import type {
  ExternalArgs,
  InternalAPI,
  ObjectBufferSettings,
} from "./interfaces";
import { ENTRY_TYPE } from "./entry-types";
import { INTERNAL_API_SYMBOL } from "./symbols";
import {
  UNDEFINED_KNOWN_ADDRESS,
  NULL_KNOWN_ADDRESS,
  TRUE_KNOWN_ADDRESS,
  FALSE_KNOWN_ADDRESS,
  ENDIANNESS,
} from "./consts";
import { FunctionalAllocatorWrapper } from "./TransactionalAllocator";

export function getEndiannessOfSystem() {
  const F64 = new Float64Array(1);
  const U32 = new Uint32Array(F64.buffer);
  F64[0] = 2;
  if (U32[1] === 0x40000000) {
    return ENDIANNESS.LITTLE;
  } else {
    return ENDIANNESS.BIG;
  }
}

export function createKnownTypeGuard<T>(arr: ReadonlyArray<T>) {
  return function knownTypeGuard(v: unknown): v is T {
    return arr.includes(v as any);
  };
}

export function invariant(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(message);
  }
}

export function arrayBufferCopyTo(
  origin: ArrayBuffer,
  startByte: number,
  length: number,
  target: ArrayBuffer,
  toTargetByte: number
) {
  const copyFrom = new Uint8Array(origin);
  const copyTo = new Uint8Array(target);

  copyTo.set(copyFrom.subarray(startByte, startByte + length), toTargetByte);
}

export function getOurPointerIfApplicable(
  value: any,
  ourAllocator: FunctionalAllocatorWrapper
) {
  if (INTERNAL_API_SYMBOL in value) {
    const api = getInternalAPI(value);
    if (api.getCarrier().allocator === ourAllocator) {
      return api.getEntryPointer();
    }
  }
}

export function externalArgsApiToExternalArgsApi(
  p: ObjectBufferSettings
): ExternalArgs {
  return {
    ...p,
    hashMapMinInitialCapacity: p.hashMapMinInitialCapacity
      ? p.hashMapMinInitialCapacity
      : 8,
    hashMapLoadFactor: p.hashMapLoadFactor ? p.hashMapLoadFactor : 0.75,
    arrayAdditionalAllocation: p.arrayAdditionalAllocation
      ? p.arrayAdditionalAllocation
      : 0,
  };
}

export function getInternalAPI(value: any): InternalAPI {
  if (!value[INTERNAL_API_SYMBOL]) {
    throw new RangeError("getInternalAPI not applicable");
  }

  return value[INTERNAL_API_SYMBOL];
}

export function align(value: number, alignTo = 8) {
  return Math.ceil(value / alignTo) * alignTo;
}

export function isKnownAddressValuePointer(entryPointer: number) {
  return (
    entryPointer === UNDEFINED_KNOWN_ADDRESS ||
    entryPointer === NULL_KNOWN_ADDRESS ||
    entryPointer === TRUE_KNOWN_ADDRESS ||
    entryPointer === FALSE_KNOWN_ADDRESS
  );
}

export function isTypeWithRC(type: ENTRY_TYPE) {
  return (
    type === ENTRY_TYPE.OBJECT ||
    type === ENTRY_TYPE.ARRAY ||
    type === ENTRY_TYPE.DATE ||
    type === ENTRY_TYPE.MAP ||
    type === ENTRY_TYPE.SET ||
    type === ENTRY_TYPE.STRING
  );
}

/**
 *
 * I hope It's reliable
 */
export function isSupportedTopLevelValue(value: unknown) {
  return !(
    Array.isArray(value) ||
    value instanceof Date ||
    value instanceof Map ||
    value instanceof Set ||
    typeof value !== "object" ||
    typeof value === null ||
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    value!.constructor.name !== "Object"
  );
}

export function fromEntries(iterable: Iterable<unknown>): unknown {
  // @ts-expect-error we don't care
  return [...iterable].reduce(function (obj: any, [key, val]) {
    obj[key] = val;
    return obj;
  }, {});
}
