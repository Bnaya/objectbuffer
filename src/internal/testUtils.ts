import { IMemPool, MemPool } from "@thi.ng/malloc";
import { OutOfMemoryError } from "./exceptions";
import { GlobalCarrier } from "./interfaces";
import { MEM_POOL_START } from "./consts";
import { createHeap } from "../structsGenerator/consts";
import { getInternalAPI } from "./utils";
import { memoryStats } from "./api";

export function getArrayBufferOnTopSize(ob: unknown) {
  const stats = memoryStats(ob);
  const carrier = getInternalAPI(ob).getCarrier();
  return carrier.heap.Uint8Array.slice(0, stats.top);
}

export function makeAllocatorThrowOnOOM(allocator: IMemPool) {
  const origMalloc = allocator.malloc;
  allocator.malloc = function wrappedMalloc(memSize) {
    if (memSize === 0) {
      throw new Error("memSize can't be 0");
    }

    const allocated = origMalloc.call(allocator, memSize);
    if (allocated === 0) {
      throw new Error("OOM memSize");
    }

    return allocated;
  };
}

export function makeThrowOnOOM(ob: unknown) {
  const allocator = getInternalAPI(ob).getCarrier().allocator;

  makeAllocatorThrowOnOOM(allocator);
}

export function arrayBuffer2HexArray(
  buffer: ArrayBuffer,
  withByteNumber = false
) {
  if (withByteNumber) {
    return Array.prototype.map.call(
      new Uint8Array(buffer),
      (x: number, index) => `${index}:${x.toString(16).padStart(4, "0x")}`
    );
  }
  return Array.prototype.map.call(new Uint8Array(buffer), (x: number) =>
    x.toString(16).padStart(4, "0x")
  );
}

export function wait(time: number) {
  return new Promise((res) => {
    // eslint-disable-next-line no-undef
    setTimeout(res, time);
  });
}

// extend pool and not monkey patch? need to think about it
export function recordAllocations(operation: () => void, pool: IMemPool) {
  const allocations = new Set<number>();
  const deallocations = new Set<number>();

  const originalMalloc = pool.malloc;
  const originalCalloc = pool.calloc;
  const originalRealloc = pool.realloc;
  const originalFree = pool.free;
  let isInReallocOrCalloc = false;

  pool.realloc = function realloc(ptr: number, size: number) {
    isInReallocOrCalloc = true;
    const allocation = originalRealloc.call(pool, ptr, size);

    if (allocation === 0) {
      throw new OutOfMemoryError();
    }

    isInReallocOrCalloc = false;

    return allocation;
  };

  pool.malloc = function malloc(size: number) {
    const allocation = originalMalloc.call(pool, size);

    if (allocation === 0) {
      throw new OutOfMemoryError();
    }

    if (!isInReallocOrCalloc) {
      allocations.add(allocation);
    }

    return allocation;
  };

  pool.calloc = function calloc(size: number) {
    isInReallocOrCalloc = true;
    const allocation = originalCalloc.call(pool, size);

    if (allocation === 0) {
      throw new OutOfMemoryError();
    }

    allocations.add(allocation);

    isInReallocOrCalloc = false;
    return allocation;
  };

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  pool.free = function free(ptr: number) {
    deallocations.add(ptr);
    allocations.delete(ptr);
    originalFree.call(this, ptr);
  };

  operation();

  pool.malloc = originalMalloc;
  pool.calloc = originalCalloc;
  pool.realloc = originalRealloc;
  pool.free = originalFree;

  return { allocations: [...allocations], deallocations: [...deallocations] };
}

export function makeCarrier(arrayBuffer: ArrayBuffer) {
  const allocator = new MemPool({
    buf: arrayBuffer,
    start: MEM_POOL_START,
  });

  const carrier: GlobalCarrier = {
    allocator,
    heap: createHeap(arrayBuffer),
  };

  return carrier;
}
