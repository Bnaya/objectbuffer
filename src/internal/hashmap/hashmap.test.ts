/* eslint-env jest */

import { arrayBuffer2HexArray, makeCarrier } from "../testUtils";
import {
  createHashMap,
  hashMapInsertUpdate,
  hashMapValueLookup,
  hashMapLowLevelIterator,
  hashMapNodePointerToKeyValue,
  hashMapSize,
  hashMapDelete,
} from "./hashmap";
import { GlobalCarrier } from "../interfaces";
import { externalArgsApiToExternalArgsApi } from "../utils";
import { readString } from "../readString";

describe("hashmap", () => {
  const externalArgs = externalArgsApiToExternalArgsApi({
    arrayAdditionalAllocation: 20,
  });

  let ab = new ArrayBuffer(128);
  let carrier: GlobalCarrier = makeCarrier(ab);

  function setABSize(size: number) {
    ab = new ArrayBuffer(size);
    carrier = makeCarrier(ab);
  }

  beforeEach(() => {
    setABSize(256);
  });

  test("createHashMap", () => {
    const mapPointer = createHashMap(carrier, 8);
    expect(mapPointer).toMatchInlineSnapshot(`56`);
    expect(
      arrayBuffer2HexArray(ab.slice(0, carrier.allocator.stats().top), true)
    ).toMatchSnapshot("simple create empty");
  });

  test("hashMapInsert simple, key is number", () => {
    const mapPointer = createHashMap(carrier, 8);

    const valuePointer = hashMapInsertUpdate(
      externalArgs,
      carrier,
      mapPointer,
      3
    );

    carrier.heap.Uint32Array[valuePointer / Uint32Array.BYTES_PER_ELEMENT] = 5;

    expect(
      arrayBuffer2HexArray(ab.slice(0, carrier.allocator.stats().top), true)
    ).toMatchSnapshot("after insert");
  });

  test("hashMapInsert simple, key is string", () => {
    const mapPointer = createHashMap(carrier, 8);

    const pointer = hashMapInsertUpdate(
      externalArgs,
      carrier,
      mapPointer,
      "abc"
    );

    carrier.heap.Uint32Array[pointer / Uint32Array.BYTES_PER_ELEMENT] = 6;

    expect(
      arrayBuffer2HexArray(ab.slice(0, carrier.allocator.stats().top), true)
    ).toMatchSnapshot("after insert");
  });

  test("hashMapInsert / hashMapValueLookup simple, key is a number", () => {
    const mapPointer = createHashMap(carrier, 8);

    const key = 3;

    const pointer = hashMapInsertUpdate(externalArgs, carrier, mapPointer, key);

    const foundValuePointer = hashMapValueLookup(carrier.heap, mapPointer, key);

    expect(pointer).toBe(foundValuePointer);
  });

  test("hashMapInsert / hashMapValueLookup simple, key is a string", () => {
    const mapPointer = createHashMap(carrier, 8);

    const key = "abc";

    const valuePointer = hashMapInsertUpdate(
      externalArgs,
      carrier,
      mapPointer,
      key
    );

    const foundValuePointer = hashMapValueLookup(carrier.heap, mapPointer, key);
    expect(foundValuePointer).toBe(valuePointer);
  });

  test("hashMapInsert / hashMapValueLookup, same key give same pointer", () => {
    const mapPointer = createHashMap(carrier, 8);

    const key = "abc";

    const firstValuePointer = hashMapInsertUpdate(
      externalArgs,
      carrier,
      mapPointer,
      key
    );

    expect(firstValuePointer).toMatchInlineSnapshot(`152`);

    const foundValuePointer = hashMapValueLookup(carrier.heap, mapPointer, key);

    expect(foundValuePointer).toBe(firstValuePointer);
  });

  test("hashMapValueLookup none existing key", () => {
    const mapPointer = createHashMap(carrier);

    const key = "abc";

    const memoryOfOverWrittenValue = hashMapInsertUpdate(
      externalArgs,
      carrier,
      mapPointer,
      key
    );

    expect(memoryOfOverWrittenValue).toMatchInlineSnapshot(`160`);

    const foundValuePointer = hashMapValueLookup(
      carrier.heap,
      mapPointer,
      "Not a real key"
    );
    expect(foundValuePointer).toBe(0);
  });

  test("hashMapLowLevelIterator", () => {
    setABSize(2048 * 2);
    const mapPointer = createHashMap(carrier);

    const input = [...new Array(26).keys()]
      .map((i): number => i + "a".charCodeAt(0))
      .map((n) => String.fromCharCode(n));

    const inserts: number[] = [];

    for (const [index, value] of input.entries()) {
      inserts.push(
        hashMapInsertUpdate(externalArgs, carrier, mapPointer, value)
      );

      carrier.heap.Uint32Array[
        inserts[inserts.length - 1] / Uint32Array.BYTES_PER_ELEMENT
      ] = index;
    }

    const values: Array<{
      valuePointer: number;
      keyPointer: number;
    }> = [];

    let iteratorToken = 0;
    while (
      (iteratorToken = hashMapLowLevelIterator(
        carrier.heap,
        mapPointer,
        iteratorToken
      )) !== 0
    ) {
      values.push(hashMapNodePointerToKeyValue(carrier.heap, iteratorToken));
    }
    expect(values.map((v) => readString(carrier.heap, v.keyPointer)))
      .toMatchInlineSnapshot(`
      Array [
        "a",
        "b",
        "c",
        "d",
        "e",
        "f",
        "g",
        "h",
        "i",
        "j",
        "k",
        "l",
        "m",
        "n",
        "o",
        "p",
        "q",
        "r",
        "s",
        "t",
        "u",
        "v",
        "w",
        "x",
        "y",
        "z",
      ]
    `);
  });

  test("hashMapSize", () => {
    const abSize = 2048 * 4;
    setABSize(abSize);
    const mapPointer = createHashMap(carrier);

    expect(hashMapSize(carrier.heap, mapPointer)).toMatchInlineSnapshot(`0`);
    const memAvailableAfterEachStep = [carrier.allocator.stats().top];

    const input = [...new Array(26).keys()]
      .map((i): number => i + "a".charCodeAt(0))
      .map((n) => String.fromCharCode(n));

    for (const [index, useThatAsKey] of input.entries()) {
      carrier.heap.Uint32Array[
        hashMapInsertUpdate(externalArgs, carrier, mapPointer, useThatAsKey) /
          Uint32Array.BYTES_PER_ELEMENT
      ] = index;

      memAvailableAfterEachStep.push(carrier.allocator.stats().top);
    }

    expect(hashMapSize(carrier.heap, mapPointer)).toBe(26);

    hashMapDelete(carrier, mapPointer, "a");
    hashMapDelete(carrier, mapPointer, "b");
    hashMapDelete(carrier, mapPointer, "c");
    hashMapDelete(carrier, mapPointer, "t");
    expect(hashMapSize(carrier.heap, mapPointer)).toBe(22);
    memAvailableAfterEachStep.push(carrier.allocator.stats().top);

    expect(memAvailableAfterEachStep).toMatchInlineSnapshot(`
      Array [
        152,
        240,
        328,
        416,
        504,
        592,
        680,
        768,
        944,
        992,
        1080,
        1168,
        1256,
        1344,
        1432,
        1520,
        1776,
        1776,
        1864,
        1952,
        2040,
        2128,
        2216,
        2304,
        2392,
        2480,
        2568,
        2568,
      ]
    `);
  });
});
