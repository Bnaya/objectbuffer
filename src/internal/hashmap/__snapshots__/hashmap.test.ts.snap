// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hashmap createHashMap: simple create empty 1`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x00",
  "5:0x00",
  "6:0x00",
  "7:0x00",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0x00",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0x78",
  "17:0x00",
  "18:0x00",
  "19:0x00",
  "20:0x88",
  "21:0x00",
  "22:0x00",
  "23:0x00",
  "24:0x00",
  "25:0x01",
  "26:0x00",
  "27:0x00",
  "28:0x08",
  "29:0x00",
  "30:0x00",
  "31:0x00",
  "32:0x03",
  "33:0x00",
  "34:0x00",
  "35:0x00",
  "36:0x10",
  "37:0x00",
  "38:0x00",
  "39:0x00",
  "40:0x18",
  "41:0x00",
  "42:0x00",
  "43:0x00",
  "44:0x00",
  "45:0x00",
  "46:0x00",
  "47:0x00",
  "48:0x48",
  "49:0x00",
  "50:0x00",
  "51:0x00",
  "52:0x70",
  "53:0x00",
  "54:0x00",
  "55:0x00",
  "56:0x00",
  "57:0x00",
  "58:0x00",
  "59:0x00",
  "60:0x08",
  "61:0x00",
  "62:0x00",
  "63:0x00",
  "64:0x28",
  "65:0x00",
  "66:0x00",
  "67:0x00",
  "68:0x28",
  "69:0x00",
  "70:0x00",
  "71:0x00",
  "72:0x00",
  "73:0x00",
  "74:0x00",
  "75:0x00",
  "76:0x00",
  "77:0x00",
  "78:0x00",
  "79:0x00",
  "80:0x00",
  "81:0x00",
  "82:0x00",
  "83:0x00",
  "84:0x00",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x00",
  "89:0x00",
  "90:0x00",
  "91:0x00",
  "92:0x00",
  "93:0x00",
  "94:0x00",
  "95:0x00",
  "96:0x00",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x00",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x10",
  "105:0x00",
  "106:0x00",
  "107:0x00",
  "108:0x40",
  "109:0x00",
  "110:0x00",
  "111:0x00",
  "112:0x80",
  "113:0x00",
  "114:0x00",
  "115:0x00",
  "116:0x80",
  "117:0x00",
  "118:0x00",
  "119:0x00",
  "120:0x10",
  "121:0x00",
  "122:0x00",
  "123:0x00",
  "124:0x68",
  "125:0x00",
  "126:0x00",
  "127:0x00",
  "128:0x00",
  "129:0x00",
  "130:0x00",
  "131:0x00",
  "132:0x00",
  "133:0x00",
  "134:0x00",
  "135:0x00",
]
`;

exports[`hashmap hashMapInsert simple, key is number: after insert 1`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x00",
  "5:0x00",
  "6:0x00",
  "7:0x00",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0x00",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0xb8",
  "17:0x00",
  "18:0x00",
  "19:0x00",
  "20:0xc8",
  "21:0x00",
  "22:0x00",
  "23:0x00",
  "24:0x00",
  "25:0x01",
  "26:0x00",
  "27:0x00",
  "28:0x08",
  "29:0x00",
  "30:0x00",
  "31:0x00",
  "32:0x03",
  "33:0x00",
  "34:0x00",
  "35:0x00",
  "36:0x10",
  "37:0x00",
  "38:0x00",
  "39:0x00",
  "40:0x18",
  "41:0x00",
  "42:0x00",
  "43:0x00",
  "44:0x00",
  "45:0x00",
  "46:0x00",
  "47:0x00",
  "48:0x48",
  "49:0x00",
  "50:0x00",
  "51:0x00",
  "52:0x70",
  "53:0x00",
  "54:0x00",
  "55:0x00",
  "56:0x01",
  "57:0x00",
  "58:0x00",
  "59:0x00",
  "60:0x08",
  "61:0x01",
  "62:0x00",
  "63:0x00",
  "64:0x28",
  "65:0x00",
  "66:0x00",
  "67:0x00",
  "68:0x28",
  "69:0x00",
  "70:0x00",
  "71:0x00",
  "72:0x90",
  "73:0x00",
  "74:0x00",
  "75:0x00",
  "76:0x00",
  "77:0x00",
  "78:0x00",
  "79:0x00",
  "80:0x00",
  "81:0x00",
  "82:0x00",
  "83:0x00",
  "84:0x00",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x00",
  "89:0x00",
  "90:0x00",
  "91:0x00",
  "92:0x00",
  "93:0x00",
  "94:0x00",
  "95:0x00",
  "96:0x00",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x00",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x10",
  "105:0x00",
  "106:0x00",
  "107:0x00",
  "108:0x40",
  "109:0x00",
  "110:0x00",
  "111:0x00",
  "112:0xc0",
  "113:0x00",
  "114:0x00",
  "115:0x00",
  "116:0x80",
  "117:0x00",
  "118:0x00",
  "119:0x00",
  "120:0x10",
  "121:0x00",
  "122:0x00",
  "123:0x00",
  "124:0x68",
  "125:0x00",
  "126:0x00",
  "127:0x00",
  "128:0xc0",
  "129:0x00",
  "130:0x00",
  "131:0x00",
  "132:0x90",
  "133:0x00",
  "134:0x00",
  "135:0x00",
  "136:0x18",
  "137:0x00",
  "138:0x00",
  "139:0x00",
  "140:0x78",
  "141:0x00",
  "142:0x00",
  "143:0x00",
  "144:0x05",
  "145:0x00",
  "146:0x00",
  "147:0x00",
  "148:0x00",
  "149:0x00",
  "150:0x00",
  "151:0x00",
  "152:0xa8",
  "153:0x00",
  "154:0x00",
  "155:0x00",
  "156:0x80",
  "157:0x00",
  "158:0x00",
  "159:0x00",
  "160:0x18",
  "161:0x00",
  "162:0x00",
  "163:0x00",
  "164:0x88",
  "165:0x00",
  "166:0x00",
  "167:0x00",
  "168:0x00",
  "169:0x00",
  "170:0x00",
  "171:0x00",
  "172:0x00",
  "173:0x00",
  "174:0x00",
  "175:0x40",
  "176:0x00",
  "177:0x00",
  "178:0x00",
  "179:0x00",
  "180:0x00",
  "181:0x00",
  "182:0x08",
  "183:0x40",
  "184:0x10",
  "185:0x00",
  "186:0x00",
  "187:0x00",
  "188:0xa0",
  "189:0x00",
  "190:0x00",
  "191:0x00",
  "192:0x00",
  "193:0x00",
  "194:0x00",
  "195:0x00",
  "196:0x00",
  "197:0x00",
  "198:0x00",
  "199:0x00",
]
`;

exports[`hashmap hashMapInsert simple, key is string: after insert 1`] = `
Array [
  "0:0x00",
  "1:0x00",
  "2:0x00",
  "3:0x00",
  "4:0x00",
  "5:0x00",
  "6:0x00",
  "7:0x00",
  "8:0x00",
  "9:0x00",
  "10:0x00",
  "11:0x00",
  "12:0x00",
  "13:0x00",
  "14:0x00",
  "15:0x00",
  "16:0xc8",
  "17:0x00",
  "18:0x00",
  "19:0x00",
  "20:0xd8",
  "21:0x00",
  "22:0x00",
  "23:0x00",
  "24:0x00",
  "25:0x01",
  "26:0x00",
  "27:0x00",
  "28:0x08",
  "29:0x00",
  "30:0x00",
  "31:0x00",
  "32:0x03",
  "33:0x00",
  "34:0x00",
  "35:0x00",
  "36:0x10",
  "37:0x00",
  "38:0x00",
  "39:0x00",
  "40:0x18",
  "41:0x00",
  "42:0x00",
  "43:0x00",
  "44:0x00",
  "45:0x00",
  "46:0x00",
  "47:0x00",
  "48:0x48",
  "49:0x00",
  "50:0x00",
  "51:0x00",
  "52:0x70",
  "53:0x00",
  "54:0x00",
  "55:0x00",
  "56:0x01",
  "57:0x00",
  "58:0x00",
  "59:0x00",
  "60:0x08",
  "61:0x01",
  "62:0x00",
  "63:0x00",
  "64:0x28",
  "65:0x00",
  "66:0x00",
  "67:0x00",
  "68:0x28",
  "69:0x00",
  "70:0x00",
  "71:0x00",
  "72:0x00",
  "73:0x00",
  "74:0x00",
  "75:0x00",
  "76:0x00",
  "77:0x00",
  "78:0x00",
  "79:0x00",
  "80:0x90",
  "81:0x00",
  "82:0x00",
  "83:0x00",
  "84:0x00",
  "85:0x00",
  "86:0x00",
  "87:0x00",
  "88:0x00",
  "89:0x00",
  "90:0x00",
  "91:0x00",
  "92:0x00",
  "93:0x00",
  "94:0x00",
  "95:0x00",
  "96:0x00",
  "97:0x00",
  "98:0x00",
  "99:0x00",
  "100:0x00",
  "101:0x00",
  "102:0x00",
  "103:0x00",
  "104:0x10",
  "105:0x00",
  "106:0x00",
  "107:0x00",
  "108:0x40",
  "109:0x00",
  "110:0x00",
  "111:0x00",
  "112:0xd0",
  "113:0x00",
  "114:0x00",
  "115:0x00",
  "116:0x80",
  "117:0x00",
  "118:0x00",
  "119:0x00",
  "120:0x10",
  "121:0x00",
  "122:0x00",
  "123:0x00",
  "124:0x68",
  "125:0x00",
  "126:0x00",
  "127:0x00",
  "128:0xd0",
  "129:0x00",
  "130:0x00",
  "131:0x00",
  "132:0x90",
  "133:0x00",
  "134:0x00",
  "135:0x00",
  "136:0x18",
  "137:0x00",
  "138:0x00",
  "139:0x00",
  "140:0x78",
  "141:0x00",
  "142:0x00",
  "143:0x00",
  "144:0x06",
  "145:0x00",
  "146:0x00",
  "147:0x00",
  "148:0x00",
  "149:0x00",
  "150:0x00",
  "151:0x00",
  "152:0xa8",
  "153:0x00",
  "154:0x00",
  "155:0x00",
  "156:0x80",
  "157:0x00",
  "158:0x00",
  "159:0x00",
  "160:0x18",
  "161:0x00",
  "162:0x00",
  "163:0x00",
  "164:0x88",
  "165:0x00",
  "166:0x00",
  "167:0x00",
  "168:0x00",
  "169:0x00",
  "170:0x00",
  "171:0x00",
  "172:0x00",
  "173:0x00",
  "174:0x14",
  "175:0x40",
  "176:0x03",
  "177:0x00",
  "178:0x00",
  "179:0x00",
  "180:0xc0",
  "181:0x00",
  "182:0x00",
  "183:0x00",
  "184:0x10",
  "185:0x00",
  "186:0x00",
  "187:0x00",
  "188:0xa0",
  "189:0x00",
  "190:0x00",
  "191:0x00",
  "192:0x61",
  "193:0x62",
  "194:0x63",
  "195:0x00",
  "196:0x00",
  "197:0x00",
  "198:0x00",
  "199:0x00",
  "200:0x10",
  "201:0x00",
  "202:0x00",
  "203:0x00",
  "204:0xb8",
  "205:0x00",
  "206:0x00",
  "207:0x00",
  "208:0x00",
  "209:0x00",
  "210:0x00",
  "211:0x00",
  "212:0x00",
  "213:0x00",
  "214:0x00",
  "215:0x00",
]
`;
